File: NewsDaily2/app/src/main/java/com/andruid/magic/newsdaily/fragment/SettingsFragment.java
package com.andruid.magic.newsdaily.fragment;

import android.os.Bundle;

import androidx.fragment.app.DialogFragment;
import androidx.preference.Preference;
import androidx.preference.PreferenceFragmentCompat;
import androidx.preference.PreferenceManager;

import com.andruid.magic.newsdaily.R;
import com.andruid.magic.newsdaily.pref.CountryPreference;
import com.andruid.magic.newsdaily.pref.CountryPreferenceDialogFragment;
import com.andruid.magic.newsdaily.util.PrefUtil;

import java.util.Objects;

public class SettingsFragment extends PreferenceFragmentCompat {

    public static SettingsFragment newInstance() {
        return new SettingsFragment();
    }

    @Override
    public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
        addPreferencesFromResource(R.xml.app_preferences);
        String defCountry = PrefUtil.getDefaultCountry(Objects.requireNonNull(getContext()));
        String country = PreferenceManager.getDefaultSharedPreferences(getContext()).getString(
                getString(R.string.pref_country), defCountry);
        CountryPreference countryPreference = findPreference(getString(R.string.pref_country));
        if(countryPreference != null)
            countryPreference.setCountry(country);
    }

    @Override
    public void onDisplayPreferenceDialog(Preference preference) {
        DialogFragment dialogFragment = null;
        if(preference instanceof CountryPreference)
            dialogFragment = CountryPreferenceDialogFragment.newInstance(preference.getKey());
        if(dialogFragment != null){
            dialogFragment.setTargetFragment(this, 0);
            dialogFragment.show(Objects.requireNonNull(getFragmentManager()),
                    getString(R.string.pref_country));
        } else
            super.onDisplayPreferenceDialog(preference);
    }
}

File: NewsDaily/app/src/main/java/com/death14stroke/newsdaily/ui/fragment/SettingsFragment.kt
package com.death14stroke.newsdaily.ui.fragment

import android.os.Bundle
import android.view.View
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import com.death14stroke.newsdaily.R
import com.death14stroke.newsdaily.data.repository.MainRepository
import com.death14stroke.newsdaily.ui.custom.CountryPreference
import com.death14stroke.newsdaily.ui.util.getColorFromAttr
import org.koin.android.ext.android.inject

class SettingsFragment : PreferenceFragmentCompat() {
    private val repository by inject<MainRepository>()
    private val countryPreference by lazy { findPreference<CountryPreference>(getString(R.string.pref_country)) }

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        addPreferencesFromResource(R.xml.app_preferences)
        val country = repository.getSelectedCountry()
        countryPreference?.country = country
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        view.setBackgroundColor(requireContext().getColorFromAttr(com.google.android.material.R.attr.colorSurface))
    }

    @Suppress("DEPRECATION")
    override fun onDisplayPreferenceDialog(preference: Preference) {
        if (preference is CountryPreference) {
            val fragment = CountryPreferenceDialogFragment.newInstance(preference.key).apply {
                setTargetFragment(this@SettingsFragment, 0)
            }
            fragment.show(parentFragmentManager, "countryPref")
        } else
            super.onDisplayPreferenceDialog(preference)
    }
}